### ISSUE TITLE
FITS int Column unexpectedly converted to MaskedColumn

### ISSUE THREAD

**Issue by taldcroft** [astropy/astropy#15417]
### Description

The round-trip behavior of a Table integer `Column` changes if there is another `MaskedColumn` in the table.

### Expected behavior

`Column` and `MaskedColumn` should be independent.

### How to Reproduce

```python
from astropy.table import Table, MaskedColumn

t1 = Table([[1, 2], [1.5, 2.5]], names=["a", "b"])
t1.write("t1.fits", overwrite=True)
t1_1 = Table.read("t1.fits")
print(t1.info)
print(t1_1.info)
```
Output:
```
<Table length=2>
name  dtype 
---- -------
   a   int64
   b float64

<Table length=2>
name  dtype 
---- -------
   a   int64
   b float64
```
Now add a Masked integer column:
```
t2 = t1.copy()
t2["c"] = MaskedColumn([1, 2], mask=[True, False])
t2.write("t2.fits", overwrite=True)
t2_1 = Table.read("t2.fits")
print(t2.info)
print(t2_1.info)
```
Output (column `a` got converted to Masked somewhere in the round-trip):
```
<Table length=2>
name  dtype     class     n_bad
---- ------- ------------ -----
   a   int64       Column     0
   b float64       Column     0
   c   int64 MaskedColumn     1

<Table length=2>
name  dtype     class     n_bad
---- ------- ------------ -----
   a   int64 MaskedColumn     0
   b float64       Column     0
   c   int64 MaskedColumn     1
```

### Versions

import platform; print(platform.platform())
import sys; print("Python", sys.version)
import astropy; print("astropy", astropy.__version__)
import numpy; print("Numpy", numpy.__version__)
import erfa; print("pyerfa", erfa.__version__)
import scipy; print("Scipy", scipy.__version__)
import matplotlib; print("Matplotlib", matplotlib.__version__)
```
macOS-13.5.2-x86_64-i386-64bit
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:27:35) [Clang 14.0.6 ]
astropy 5.3.1
Numpy 1.23.5
pyerfa 2.0.0.1
Scipy 1.10.0
Matplotlib 3.6.3
```

**Comment by saimn** [comment]
It's because the `Table` is converted first to a Numpy (masked) recarray:
https://github.com/astropy/astropy/blob/main/astropy/io/fits/convenience.py#L512-L513

which doesn't keep track of which columns were masked and just use the same fill value for a given dtype:
```
(Pdb++) tarray
masked_array(data=[(1, 1.5, --), (2, 2.5, 2)],
             mask=[(False, False,  True), (False, False, False)],
       fill_value=(999999, nan, 999999),
            dtype=[('a', '<i8'), ('b', '<f8'), ('c', '<i8')])
(Pdb++) tarray.fill_value
(999999, nan, 999999)
```

cc @mhvk since you added `table.as_array()` in #11222 ;).

**Comment by mhvk** [comment]
@saimn - ah, yes, that definitely is the problem. Looks like in the loop right after one should check whether the original column is actually masked before setting `col.null`: https://github.com/astropy/astropy/blob/e16a7b887aa9592186f467604dc9555291ee801c/astropy/io/fits/convenience.py#L540-L549

**Comment by mhvk** [comment]
For a fix, see #15473

**Comment by taldcroft** [comment]
Thanks @mhvk!!
